release_notes: append #PR to markdown link text

scpeters opened the pull request:
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [X] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

I noticed in the [release notes for 3.1.4](https://github.com/Homebrew/brew/releases/tag/3.1.4) that some lines have an author name but no pull request. As of this writing, the release notes start with the following markdown code (reproduce with `brew ruby -rrelease_notes -e 'puts ReleaseNotes.generate_release_notes("3.1.3", "3.1.4", markdown: true)'`):

~~~
- [upgrade: fix broken dependents never being built from source](https://github.com/Homebrew/brew/pull/11263) (@Bo98)
- [Update maintainers, manpage and completions.](https://github.com/Homebrew/brew/pull/11261) (@Homebrew)
- [Allow man and prof Apple Silicon](https://github.com/Homebrew/brew/pull/11256) (@dtrodrigues)
- [Interesting-Taps-and-Forks.md: add gromgit/fuse](https://github.com/Homebrew/brew/pull/11255) (@gromgit)
- [docs: add Cask Cookbook](https://github.com/Homebrew/brew/pull/11249) (@SMillerDev)
- [dev-cmd/bottle: use gnu-tar universally, sort entries & use PAX](https://github.com/Homebrew/brew/pull/11253) (@Bo98)
- [](https://github.com/Homebrew/brew/pull/11129) (@SMillerDev)
~~~

Notice that the last line has a markdown link to #11129, but it is not rendered because it has no link text. I believe this is because the merge commit for that pull request https://github.com/Homebrew/brew/commit/afa99b49630614119321429c8e84b37941bd228d has only one line of description, while other merge commits, such as https://github.com/Homebrew/brew/commit/94ce3236a517dfae6b492dcb4fd6b726a6175e8f for https://github.com/Homebrew/brew/pull/11263, have an additional line of description.

There are 3 pull requests in 3.1.4 that have this issue. I don't know why these merge commits are like this. As a workaround, I've modified `release_notes.rb` to append an extra ` #11129` to the Markdown link text. With this change the beginning of the 3.1.4 release notes would have the following markdown:

~~~
- [upgrade: fix broken dependents never being built from source #11263](https://github.com/Homebrew/brew/pull/11263) (@Bo98)
- [Update maintainers, manpage and completions. #11261](https://github.com/Homebrew/brew/pull/11261) (@Homebrew)
- [Allow man and prof Apple Silicon #11256](https://github.com/Homebrew/brew/pull/11256) (@dtrodrigues)
- [Interesting-Taps-and-Forks.md: add gromgit/fuse #11255](https://github.com/Homebrew/brew/pull/11255) (@gromgit)
- [docs: add Cask Cookbook #11249](https://github.com/Homebrew/brew/pull/11249) (@SMillerDev)
- [dev-cmd/bottle: use gnu-tar universally, sort entries & use PAX #11253](https://github.com/Homebrew/brew/pull/11253) (@Bo98)
- [ #11129](https://github.com/Homebrew/brew/pull/11129) (@SMillerDev)
~~~

Rylan12 commented on Library/Homebrew/release_notes.rb:
```diff
@@ -25,8 +25,12 @@ def generate_release_notes(start_ref, end_ref, markdown: false)
+        # (https://github.com/Homebrew/pull)/(1234) (@author) - (title)
+        # maps to
+        # [title #1234](https://github.com/Homebrew/pull/1234) (@author)
+        %r{(.*)/(\d+) \(@(.+)\) - (.*)} =~ s
+        "- [#{Regexp.last_match(4)} \##{Regexp.last_match(2)}](#{Regexp.last_match(1)}/#{Regexp.last_match(2)})" \
+        " (@#{Regexp.last_match(3)})\n"
```
```suggestion
        matches =~ s.match %r{(?<url>.*)/(?<pr>\d+) \(@(?<user>.+)\) - (?<title>.*)}
        "- [#{matches[:title]} \##{matches[:pr]}](#{matches[:url]}/#{matches[:pr]})" (@#{matches[:user]})\n"
```

I think using named captures will help readability here

scpeters replied:
your suggestions sound good to me; I'm about to go to bed, so please feel free to push to this branch as you see fit

Rylan12 replied:
> I'm about to go to bed

Same here (should have done that hours ago) 🛌😴💤

I'll take a look when I'm available tomorrow if you haven't had a chance yet.

Rylan12 reviewed (commented):
Thanks for looking into this. When merging, the merger does have the option to modify the commit message I believe so it's possible that the people who merged removed that line that is missing. Although, I was the one who merged one of the missing ones and I'm almost positive I didn't touch the merge commit message so it's possible that there's a reason that line's being omitted that isn't coming to me.

I like this solution, though. I wonder, though, if it looks better to use this instead:
> - [#11191:](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
> - [#11192: workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)

Or maybe bold (with or without the colon):

> - [**#11191**:](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
> - [**#11192**: workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)

> - [**#11191**](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
> - [**#11192** workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)

That way the PR number is always first so if the title is missing it's a little more consistent. Plus, it might make a little more sense to have the PR number first anyway.

Also, I don't think we need to rely on a single string interpolation to get this right. If we need to use some more complex logic to change the markdown line if the title is blank we definitely have the room to do so. It might also make it cleaner-looking.

MikeMcQuaid commented:
> Notice that the last line has a markdown link to #11129, but it is not rendered because it has no link text. I believe this is because the merge commit for that pull request [afa99b4](https://github.com/Homebrew/brew/commit/afa99b49630614119321429c8e84b37941bd228d) has only one line of description, while other merge commits, such as [94ce323](https://github.com/Homebrew/brew/commit/94ce3236a517dfae6b492dcb4fd6b726a6175e8f) for #11263, have an additional line of description.

I'd suggest we consider falling back to the first line of the merge commit if the second line isn't present. That at least provides a link to the PR and a branch name that may provide more context.

scpeters commented:
> > Notice that the last line has a markdown link to #11129, but it is not rendered because it has no link text. I believe this is because the merge commit for that pull request [afa99b4](https://github.com/Homebrew/brew/commit/afa99b49630614119321429c8e84b37941bd228d) has only one line of description, while other merge commits, such as [94ce323](https://github.com/Homebrew/brew/commit/94ce3236a517dfae6b492dcb4fd6b726a6175e8f) for #11263, have an additional line of description.
> 
> I'd suggest we consider falling back to the first line of the merge commit if the second line isn't present. That at least provides a link to the PR and a branch name that may provide more context.

so it would look like this:

- [Merge pull request #11129 from SMillerDev/feature/audit/service_command](https://github.com/Homebrew/brew/pull/11129) (@SMillerDev)

scpeters commented:
to summarize some of the proposed display options:

this PR as of 6e91c35:

* [ #11191](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
* [workflows/tests: simulate macOS for `brew style homebrew-core` #11192](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)

from @Rylan12:

> > * [#11191:](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
> > * [#11192: workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)
> 
> Or maybe bold (with or without the colon):
> 
> > * [**#11191**:](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
> > * [**#11192**: workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)
> 
> > * [**#11191**](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
> > * [**#11192** workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)

from @MikeMcQuaid 

* [Merge pull request #11191 from Rylan12/re-enable-gitlab-patch-check](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
* [workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)

scpeters commented:
here's a blend of those proposals that I like:

* [**#11191**: Merge pull request #11191 from Rylan12/re-enable-gitlab-patch-check](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
* [**#11192**: workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)

MikeMcQuaid commented:
Personally I think displaying the PR number is redundant if it's going to be link text.

Another option that occurred to me that'd be more involved would be to extract the PR number and query the title from the GitHub API.

Rylan12 commented:
Yeah, I think I agree that including the number just adds extra text that isn't really that useful if there's already a link. I guess we could list the PR only if the description is blank:

> * [#11191](https://github.com/Homebrew/brew/pull/11191) (@Rylan12)
> * [workflows/tests: simulate macOS for `brew style homebrew-core`](https://github.com/Homebrew/brew/pull/11192) (@Rylan12)

> Another option that occurred to me that'd be more involved would be to extract the PR number and query the title from the GitHub API.

Yeah, I wouldn't be opposed to this. We already have the number so it shouldn't be that much more work. Plus it doesn't seem to happen that often so doing an API request every once in a while when `brew release` is run seems fine. If we really wanted we could base the release notes purely off of the titles from the API but that's probably an unnecessary number of queries for the same result.

scpeters commented:
I just added a new test case in fe994a3 for the behavior when a merge commit has a subject message but no body message. Let's use that new test expectation to capture the desired formatting

scpeters commented:
I made a quick attempt at updating the regular expression code, but I had some trouble. I won't be able to work on this for a few days.

MikeMcQuaid commented on Library/Homebrew/test/release_notes_spec.rb:
```diff
@@ -12,21 +12,27 @@
 
   describe ".generate_release_notes" do
     it "generates release notes" do
+      # rubocop:disable Style/TrailingWhitespace
```
Seems undesirable to have trailing whitespace generated here?

scpeters replied:
that was a placeholder, and it's gone now

MikeMcQuaid commented on Library/Homebrew/test/release_notes_spec.rb:
```diff
@@ -12,21 +12,27 @@
       expect(described_class.generate_release_notes("release-notes-testing", "HEAD", markdown: true)).to eq <<~NOTES
-        - [Do something else](https://github.com/Homebrew/brew/pull/2) (@User)
-        - [Do something](https://github.com/Homebrew/brew/pull/1) (@Homebrew)
+        - [ #3](https://github.com/Homebrew/brew/pull/3) (@User)
```
Similarly this output seems a bit weird. How about `Merge pull request #3 from User/another_change` or anything relating to `another_change` (as it's a branch name)?

scpeters replied:
I've used the subject line of the commit if the body is empty (so `Merge pull request #3 from User/another_change`)

scpeters commented:
I reverted my changes with the PR number but kept the new test case and added code based on @MikeMcQuaid's suggestion to use the subject line of the commit message if the body is missing. I'll leave the approach using the GitHub API for another day.

I rebased to clean this up but the commits aren't verified. I tried following the gpg signing guide on my Linux machine, but it's not currently working (my previous commits were verified because I committed through the web browser). I can try to configure my mac with gpg signing and re-push if needed

MikeMcQuaid reviewed (approved):
Looks great, nice work here!
